gemfire-core/src/main/java/com/gemstone/gemfire/internal/DirectByteBufferDataOutput.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2017 SnappyData, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+
+package com.gemstone.gemfire.internal;
+
+import java.io.Closeable;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.io.UTFDataFormatException;
+import java.nio.ByteBuffer;
+import javax.annotation.Nonnull;
+
+import com.esotericsoftware.kryo.KryoException;
+import com.esotericsoftware.kryo.io.ByteBufferOutput;
+import com.gemstone.gemfire.internal.shared.ClientSharedUtils;
+import com.gemstone.gemfire.internal.shared.Version;
+import com.gemstone.gemfire.internal.shared.unsafe.ChannelBufferUnsafeDataOutputStream;
+import com.gemstone.gemfire.internal.shared.unsafe.UnsafeHolder;
+
+/**
+ * Implements {@link DataOutput} writing to a direct ByteBuffer
+ * expanding it as required.
+ * <p>
+ * Note: this can be further optimized by using the Unsafe API rather than
+ * going through ByteBuffer API (e.g. see ChannelBufferUnsafeDataOutputStream)
+ * but won't have an effect for large byte array writes (like for column data)
+ */
+public final class DirectByteBufferDataOutput extends ByteBufferOutput
+    implements DataOutput, Closeable, VersionedDataStream {
+
+  private static final int INITIAL_SIZE = 1024;
+
+  private final Version version;
+
+  public DirectByteBufferDataOutput(Version version) {
+    this(INITIAL_SIZE, version);
+  }
+
+  public DirectByteBufferDataOutput(int initialSize, Version version) {
+    super(initialSize, Integer.MAX_VALUE); // no max limit
+    this.version = version;
+  }
+
+  @Override
+  public Version getVersion() {
+    return this.version;
+  }
+
+  @Override
+  public void writeChar(int v) throws IOException {
+    super.writeChar((char)v);
+  }
+
+  @Override
+  public void writeUTF(@Nonnull String s) throws IOException {
+    // kryo's writeString is not DataOutput UTF8 compatible
+
+    // first calculate the UTF encoded length
+    final int strLen = s.length();
+    final int utfLen = ClientSharedUtils.getUTFLength(s, strLen);
+    if (utfLen > 65535) {
+      throw new UTFDataFormatException("encoded string too long: " + utfLen
```

Comment:
This is a limitation of DataOutput.writeUTF format that writes the length as a short: https://docs.oracle.com/javase/7/docs/api/java/io/DataOutput.html#writeUTF(java.lang.String). GemFire/GemXD code that intends to write long strings uses DataSerializer.writeString